checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1705 @@
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.util.JavaExpressionParseUtil.JavaExpressionParseException;
+import org.checkerframework.framework.util.StringToJavaExpression;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.Pair;
+import org.checkerframework.javacutil.TreePathUtil;
+import org.checkerframework.javacutil.TreeUtils;
+import org.checkerframework.javacutil.TypesUtils;
+import org.plumelib.util.StringsPlume;
+
+/**
+ * An analyzer that checks consistency of {@code @MustCall} and {@code @CalledMethods} types,
+ * thereby detecting resource leaks. For any expression <em>e</em>, the analyzer ensures that when
+ * <em>e</em> goes out of scope, there exists a resource alias <em>r</em> of <em>e</em> (which might
+ * be <em>e</em> itself) such that MustCall(r) is contained in CalledMethods(r). For any <em>e</em>
+ * for which this property does not hold, the analyzer reports a {@code
+ * "required.method.not.called"} error, indicating a possible resource leak.
+ *
+ * <p>Mechanically, the analysis tracks dataflow facts about the obligations of sets of
+ * resource-aliases that refer to the same resource, and checks their must-call and called-methods
+ * types when the last reference to those sets goes out of scope. That is, this class implements a
+ * lightweight alias analysis that tracks must-alias sets for resources.
+ *
+ * <p>Throughout this class, variables named "obligation" or "obligations" are dataflow facts of
+ * type {@code ImmutableSet<LocalVarWithTree>}, each representing a set of resource aliases for some
+ * value with a non-empty {@code @MustCall} obligation. These obligations can be resolved either via
+ * ownership transfer (e.g. by being assigned into an owning field) or via their must-call
+ * obligations being contained in their called-methods type when the last reference in a set goes
+ * out of scope.
+ *
+ * <p>The algorithm here adds, modifies, or removes obligations from those it is tracking when
+ * certain code patterns are encountered. For example, a new obligation is added to the tracked set
+ * when a constructor or a method with an owning return is invoked; an obligation is modified when
+ * an expression with a tracked obligation is assigned to a local variable or a resource-alias
+ * method or constructor is called (the new local or the result of the resource-alias
+ * method/constructor is added to the existing resource alias set); an obligation can be removed
+ * when a member of a resource-alias set is passed to a method in a parameter location that is
+ * annotated as {@code @Owning} or assigned to an owning field. (The above list is not exhaustive.)
+ *
+ * <p>Throughout, this class uses the temporary-variable facilities provided by the Must Call and
```

Comment:
For what expressions are temporary variables created?  All expressions?  All expressions that are subexpressions of  a larger expression?  Something else?